### ISSUE TITLE
Assign dictionary syntax

### ISSUE THREAD

**Issue by rovinellimarco** [facebookresearch/hydra#1896]
# 🚀 Feature Request

Dictionaries are always merged. It would be nice to have an "override" keyword to put in front of a parameter to signal that we want to erase that dictionary before merging our new parameters.

Example:
````
config.yaml

model:
  batch_size: 8
  loss:
    _target_: MyCustomLoss
    from_logits: true
    alpha: 0.2
````

````
override.yaml

model:
  loss:
    _target_: NewLoss
    beta: 0.3
````

At the moment it's not possible to "override" `config.yaml` with `override.yaml` since the new `loss` dictionary would be merged and not assigned.
It would be nice to be able to write:
````
override.yaml

model:
  override loss:
    _target_: NewLoss
    beta: 0.3
````
and have as output:
```
model:
  batch_size: 8
  loss:
    _target_: NewLoss
    beta: 0.3
```
## Motivation

Not having the possibility to assign a dictionary rather than merge it makes it really hard to have defaults for configurations that can be instantiated. This is because the override of the default would simply add the new parameters and not remove the ones that cannot be used by the new class.

**Comment by Jasha10** [comment]
Hi @rovinellimarco, thank you for the feature request.
Yes, you are right that dictionaries are merged (as opposed to overwriting), and there is currently no way to change the strategy that is used.

There is a rather awkward workaround that can be used, which is to first override `model.loss` to be a null value, and then subsequently override it to be the `NewLoss` value. Though such an approach is possible, I do not recommend it.

What I'd recommend is to refactor your config files so that you can use Hydra's support for [config groups](https://hydra.cc/docs/advanced/terminology#config-group) to switch between `MyCustomLoss` and `NewLoss`.

For example:
```python
# my_app.py
from omegaconf import DictConfig, OmegaConf
import hydra

@hydra.main(config_path="conf", config_name="config")
def my_app(cfg: DictConfig) -> None:
    print(OmegaConf.to_yaml(cfg))

if __name__ == "__main__":
    my_app()
```
```yaml
# conf/config.yaml
defaults:
  - loss@model: my_custom_loss
  - _self_

model:
  batch_size: 8
```
```yaml
# conf/loss/my_custom_loss.yaml
loss:
  _target_: MyCustomLoss
  from_logits: true
  alpha: 0.2
```
```yaml
# conf/loss/new_loss.yaml
loss:
  _target_: NewLoss
  beta: 0.3
```
Here is what happens when you run the app with default settings:
```text
 $ p3 my_app.py
model:
  loss:
    _target_: MyCustomLoss
    from_logits: true
    alpha: 0.2
  batch_size: 8
```
Here is what happens when you override to use `new_loss.yaml` instead of `my_custom_loss.yaml`:
```text
$ p3 my_app.py loss@model=new_loss
model:
  loss:
    _target_: NewLoss
    beta: 0.3
  batch_size: 8
```

**Comment by jieru-hu** [comment]
thanks @Jasha10 !
@rovinellimarco 
alternatively, depending on how your config is structured, you  can consider  setting `loss: null` for overriding, something like the following would work

### config.yaml
```
model:
  batch_size: 8
  loss: null

```

### override.yaml

```
defaults:
  - config
  - _self_

model:
  loss:
    _target_: NewLoss
    beta: 0.3
```

### my_app
```
@hydra.main(config_path="conf", config_name="override")
def my_app(cfg: DictConfig) -> None:
    print(OmegaConf.to_yaml(cfg))


if __name__ == "__main__":
    my_app()
```

### output

```
model:
  batch_size: 8
  loss:
    _target_: NewLoss
    beta: 0.3

```

**Comment by Butanium** [comment]
Hey I'm new to hydra, I have a default config with all my defaiult args for my experiment. In particular all my setups
```yaml
setups:
  all_ambiguous:  # useful to see if the model can solve the ambiguous task
    prompt_from: "ambiguous"
    few_shot_from: "ambiguous"
    fs_label_from: null
  ambiguous_shots:  # useful to check whether the model usually choose between the two tasks
    prompt_from: "non_ambiguous"
    few_shot_from: "ambiguous"
    fs_label_from: null
  task_A_shots:  # check the model's ability to solve task A
    prompt_from: "mixed"
    few_shot_from: "mixed"
    fs_label_from: "A"
  task_B_shots:  # check the model's ability to solve task B
    prompt_from: "mixed"
    few_shot_from: "mixed"
    fs_label_from: "B"
```
I want to have a `debug` config that overwrite setups to only use all_ambiguous, what's the right way to do this?

**Comment by omry** [comment]
Hydra already supports replacement for the component-selection use case described here by leaving the target node missing and populating it through a config group:

```yaml
defaults:
  - loss@model.loss: my_custom_loss

model:
  batch_size: 8
  loss: ???
```

Overriding the `loss` group selects one complete config for `model.loss` before composition, so there is no previous dictionary at that node to merge with the selected value.

More general per-field or per-subtree control over replace-versus-merge semantics belongs in OmegaConf and is tracked in [OmegaConf #1312](https://github.com/omry/omegaconf/issues/1312). Hydra can take advantage of that capability once the underlying merge-policy design exists.

Closing this in favor of the supported config-group pattern and the upstream OmegaConf work.

**Comment by omry** [comment]
The broader Hydra-side design remains tracked in [#2227](https://github.com/facebookresearch/hydra/issues/2227), which is blocked on the merge-policy work in [OmegaConf #1312](https://github.com/omry/omegaconf/issues/1312).
